devicetypes/vision-nameseca/zp3111jp-4-in-1-motion.src/zp3111jp-4-in-1-motion.groovy

```diff
@@ -0,0 +1,925 @@
+}
+private getHumidityTriggerParam() {
+	return createConfigParamMap(3, "Humidity Change Trigger [1-50]\n(1% - 50%)", 1, "humidityTrigger", "1..50", 10)
+}
+private getLightTriggerParam() {
+	return createConfigParamMap(4, "Light Change Trigger [0,5-50]\n(OFF,5% - 50%)", 1, "lightTrigger", "0..50", 10)
+}
+private getMotionTimeParam() {
+	return createConfigParamMap(5, "Motion Retrigger Time [1-255]\n(1 Minute - 255 Minutes)", 1, "motionTime", "1..255", 3)
+}
+private getMotionSensitivityParam() {
+	return createConfigParamMap(6, "Motion Sensitivity [1-7]\n(1 = Most Sensitive)(7 = Least Sensitive)", 1, "motionSensitivity", "1..7", 4)
+}
+private getLedIndicatorModeParam() {
+	return createConfigParamMap(7, "LED Indicator Mode [1-3]${getNameValueSettingDesc(ledIndicatorModes)}", 1, "ledIndicatorMode", "1..3", 3)
+}
+private getParamStoredVal(param) {
+	return state["configVal${param.num}"]
+}
+private createConfigParamMap(num, name, size, prefName, range, val) {
+	if (settings?."${prefName}" != null) {
+		val = settings?."${prefName}"
+	}
+	return [
+			num: num,
+			name: name,
+			size: size,
+			prefName: prefName,
+			range: range,
+			val: val
+	]
+}
+def parse(String description) {
+	def result = []
+
+	sendLastCheckinEvent()
+
+	def cmd = zwave.parse(description, commandClassVersions)
+	if (cmd) {
+		result += zwaveEvent(cmd)
+	}
+	else {
+		logDebug "Unable to parse description: $description"
+	}
+	return result
+}
+
+def zwaveEvent(physicalgraph.zwave.commands.securityv1.SecurityMessageEncapsulation cmd) {
+	def encapCmd = cmd.encapsulatedCommand(getCommandClassVersions())
+
+	def result = []
+	if (encapCmd) {
+		result += zwaveEvent(encapCmd)
+	}
+	else {
+		log.warn "Unable to extract encapsulated cmd from $cmd"
+	}
+	return result
+}
+private getCommandClassVersions() {
+	[
+			0x20: 1,  // Basic
+			0x31: 5,	// Sensor Multilevel (v7)
+			0x59: 1,  // AssociationGrpInfo
+			0x5A: 1,  // DeviceResetLocally
+			0x5E: 2,  // ZwaveplusInfo
+			0x70: 2,  // Configuration
+			0x71: 3,  // Alarm v1 or Notification v4
+			0x72: 2,  // ManufacturerSpecific
+			0x73: 1,  // Powerlevel
+			0x7A: 2,  // FirmwareUpdateMd
+			0x80: 1,  // Battery
+			0x84: 2,  // WakeUp
+			0x85: 2,  // Association
+			0x86: 1,	// Version (2)
+			0x98: 1		// Security
+	]
+}
+def zwaveEvent(physicalgraph.zwave.commands.wakeupv2.WakeUpIntervalReport cmd) {
+	state.checkinInterval = cmd.seconds
+
+	sendUpdatingEvent()
+
+	def msg = "Minimum Check-in Interval is ${cmd.seconds / 60} Minutes"
+	if (cmd.seconds == 600) {
+		log.warn "$msg"
+	}
+	else {
+		logDebug "$msg"
+	}
+
+	runIn(5, finalizeConfiguration)
+
+	return [ createCheckIntervalEvent(cmd.seconds) ]
+}
+def zwaveEvent(physicalgraph.zwave.commands.wakeupv2.WakeUpNotification cmd) {
+	logTrace "WakeUpNotification: $cmd"
+	def cmds = []
+
+	logDebug "Device Woke Up"
+
+	if (getAttrValue("firmwareVersion")) {
+		cmds += configure()
+
+		if (cmds) {
+			cmds << "delay 2000"
```

Comment:
I would probably increase the 2000 delay to something like 15 seconds. The reason I say this is because if the configure happens during a time where the hub is also communicating with other nodes specially when doing S0,S2, or FLiRS, we could potentially run into longer delays before the `configure()` messages are sent. 

Keep in mind the delay is only applied when the messages are sent to the hub, so if you send a hub at `t = 0` and you apply a 2 second delay, there is no guarantee that the hub is going to send the first at `t~0`, it could potentially get the queued as I mentioned above, so when 2 seconds later you send the message to the hub, there is a potential the the first message has not been sent yet, so there will not be a delay between them. 

And given that `wakeUpNoMoreInfoCmd()` is important to be sent once we have received all the reports, I feel like larger delay is important here.

Reply:
I have modified the code.
Thank you for very detailed explanation and suggestions.